### PR TITLE
Add fluentAPI interfaces

### PR DESCRIPTION
**Pull request by junchuanwang**
- Moved parameters class, associate keys to the interface
- Change all tests to use interface instead
- Tested subresource generation from parent resource interface

### CODE REVIEW

File: restli-tools/src/main/resources/macros/library.vm [hunk 1]
```diff
@@ -13,6 +13,15 @@
...
    limitations under the License.
 *#
+#macro(parseMethodsFromTemplates)
+  ## Action methods
+  #parse("apiVmTemplates/action.vm")
+  ## CRUD methods
+  #foreach($method in $spec.restMethods)
+      #parse("apiVmTemplates/rest.${method.method}.vm")
+  #end
+#end
```
Comment (karthikbalasub) [inline restli-tools/src/main/resources/macros/library.vm:23]: Need to include finders and batch finders if the other PR goes first.

File: restli-tools/src/main/resources/apiVmTemplates/resource_interface.vm [hunk 2]
```diff
@@ -33,14 +31,31 @@ import java.util.Map;
...
+#end
+
+## recursively build subresource interfaces
+## TODO: should correct the indentation
 #foreach($subSpec in $spec.childSubResourceSpecs)
 #**##subResourceInterface($subSpec)##
 #end
 
+## 1. RestMethod file compile and re-organize
+## 2. propagate assocations key
```
Comment (karthikbalasub) [inline restli-tools/src/main/resources/apiVmTemplates/resource_interface.vm:59]: What do these comments mean?

Comment (junchuanwang) [reply restli-tools/src/main/resources/apiVmTemplates/resource_interface.vm:59]: They are todo items. Deleted.

File: restli-tools/src/main/resources/macros/library.vm (hunk 1 above)
Comment (junchuanwang) [reply restli-tools/src/main/resources/macros/library.vm:23]: I have reviewed the other Finder PR.
I recommend this PR go first.
Then I will help resolving conflicts for that PR.
Then I will do another patch to add interface for finders.

**Review by karthikbalasub — APPROVED** [review]

File: restli-tools/src/main/resources/apiVmTemplates/action.vm [hunk 3]
```diff
@@ -13,7 +13,55 @@ Copyright (c) 2021 LinkedIn Corp.
...
+      $spec.keyClassName $spec.idName,
+      #end
+        #set($paramLen = ${method.getRequiredParameters().size()})
+        #if($paramLen > 0)
+          #set($lastParam = ${method.getRequiredParameters().get($paramLen - 1)})
+          #if($paramLen > 1)
+            #set($endIdx= ${paramLen} - 2)
+            #foreach($idx in [0..$endIdx])
+              #set($param = ${method.getRequiredParameters().get($idx)})
+            $param.paramClassDisplayName $param.paramName,
```
Comment (karthikbalasub) [inline restli-tools/src/main/resources/apiVmTemplates/action.vm:35]: use $forEach.hasNext to simplify this :)

Comment (junchuanwang) [reply restli-tools/src/main/resources/apiVmTemplates/action.vm:35]: sure. Already did

**Pull request merged**